cpp/include/cugraph/mtmg/graph.hpp

```diff
@@ -0,0 +1,136 @@
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cugraph/mtmg/detail/device_shared_wrapper.hpp>
+#include <cugraph/mtmg/edge_property.hpp>
+#include <cugraph/mtmg/graph_view.hpp>
+#include <cugraph/mtmg/handle.hpp>
+#include <cugraph/mtmg/renumber_map.hpp>
+
+namespace cugraph {
+namespace mtmg {
+
+/**
+ * @brief Graph object for each GPU
+ */
+template <typename vertex_t, typename edge_t, bool store_transposed, bool multi_gpu>
+class graph_t : public detail::device_shared_wrapper_t<
+                  cugraph::graph_t<vertex_t, edge_t, store_transposed, multi_gpu>> {
+  using parent_t = detail::device_shared_wrapper_t<
+    cugraph::graph_t<vertex_t, edge_t, store_transposed, multi_gpu>>;
+
+ public:
+  /**
+   * @brief Create an MTMG graph view (read only)
+   */
+  auto view()
+  {
+    std::lock_guard<std::mutex> lock(parent_t::lock_);
+
+    cugraph::mtmg::graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu> result;
+
+    std::for_each(parent_t::objects_.begin(), parent_t::objects_.end(), [&result](auto& p) {
+      result.set(p.first, std::move(p.second.view()));
+    });
+
+    return result;
+  }
+};
+
+/**
+ * @brief Create an MTMG graph from an edgelist
+ *
+ * @param[in]  handle             Resource handle
+ * @param[in]  edgelist           Edgelist
+ * @param[in]  graph_properties   Graph properties
+ * @param[in]  renumber           If true, renumber graph (must be true for MG)
+ * @param[out] graph              MTMG graph is stored here
+ * @param[out] edge_weights       MTMG edge weights is stored here
+ * @param[out] edge_ids           MTMG edge ids is stored here
+ * @param[out] edge_types         MTMG edge types is stored here
+ * @param[in]  renumber_map       MTMG renumber_map is stored here
+ * @param[in]  do_expensive_check A flag to run expensive checks for input arguments (if set to
+ * `true`).
+ */
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          typename edge_id_t,
+          typename edge_type_t,
+          bool store_transposed,
+          bool multi_gpu>
+void create_graph_from_edgelist(
+  handle_t const& handle,
+  cugraph::mtmg::edgelist_t<vertex_t, weight_t, edge_id_t, edge_type_t>& edgelist,
+  graph_properties_t graph_properties,
+  bool renumber,
+  cugraph::mtmg::graph_t<vertex_t, edge_t, store_transposed, multi_gpu>& graph,
+  std::optional<cugraph::mtmg::edge_property_t<
+    cugraph::mtmg::graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu>,
+    weight_t>>& edge_weights,
+  std::optional<cugraph::mtmg::edge_property_t<
+    cugraph::mtmg::graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu>,
+    edge_id_t>>& edge_ids,
+  std::optional<cugraph::mtmg::edge_property_t<
+    cugraph::mtmg::graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu>,
+    edge_type_t>>& edge_types,
+  std::optional<cugraph::mtmg::renumber_map_t<vertex_t>>& renumber_map,
```

Comment:
Why are you taking references instead of returning?

Reply:
This function is designed to be called by multiple threads concurrently.  The implementation will insert results into the references passed in.  The only way I can think to return would be to push the host threading logic inside this function, but I was trying to allow the calling code to manage the threads.